components/camera/client.go

```diff
@@ -228,6 +228,10 @@ func (c *client) Stream(
 	return stream, nil
 }
 
+func (c *client) GetImage(ctx context.Context) (image.Image, func(), error) {
+	return c.Read(ctx)
```

Comment:
Could we `Read` and handle release so we do not need to include release func in returns? (perhaps that would make sense in the videosourcewrapper layer). 
As mentioned, this would be a larger departure from Stream.Next which assuming will cause major problems.

Reply:
I'm gonna try to get rid of release in camera components and see what happens

Reply:
Looks like we just need to move the rimage.Encode step one level in i.e. outBytes should come out of the GetImage call instead of being handled in the data collector or camera server. So release cannot be fully gotten rid of, but just handled more nested in our abstraction

Reply:
This is my WIP in `server.go`

```
	switch castedCam := cam.(type) {
	case ReadImager:
		// RSDK-8663: If available, call a method that reads exactly one image. The default
		// `ReadImage` implementation will otherwise create a gostream `Stream`, call `Next` and
		// `Close` the stream. However, between `Next` and `Close`, the stream may have pulled a
		// second image from the underlying camera. This is particularly noticeable on camera
		// clients. Where a second `GetImage` request can be processed/returned over the
		// network. Just to be discarded.
		// RSDK-9132(sean yu): In addition to what Dan said above, ReadImager is important
		// for camera components that rely on the `release` functionality provided by gostream's `Read`
		// such as viamrtsp.
		// (check that this comment is 100% true before code review then delete this paranthetical statement)
		img, release, err := castedCam.Read(ctx)
		if err != nil {
			return nil, err
		}
		defer func() {
			if release != nil {
				release()
			}
		}()
	
		actualMIME, _ := utils.CheckLazyMIMEType(req.MimeType)
		resp.MimeType = actualMIME
		outBytes, err := rimage.EncodeImage(ctx, img, req.MimeType)
		if err != nil {
			return nil, err
		}
		resp.Image = outBytes
	default:
		imgBytes, mimeType, err := cam.Image(ctx, req.MimeType, ext)
		if err != nil {
			return nil, err
		}
		actualMIME, _ := utils.CheckLazyMIMEType(mimeType)
		resp.MimeType = actualMIME
		resp.Image = imgBytes
	}
```

So I think yes, in the default case we don't encode anymore

Reply:
Ok cool I like that.

I am little confused on why we still need the `ReadImager` path here. Shouldnt the camera interface now always have `Image` defined so we can just use that?

Reply:
My thinking here is since `viamrtsp` uses the `VideoReaderFunc` and `Read`'s `release` functionality to keep track of pool frames in the frame allocation optimization flow, for the `server.go` that serves `viamrtsp`, we need to be able to call `release()`

I think we could refactor `viamrtsp` though to just copy out the bytes on return early and use the new `Image` pathway... I'm down to remove `ReadImager` entirely as long as we make it a high priority to refactor `viamrtsp` to use `Image` and `[]byte`

Reply:
I see, interesting. I think it makes sense to leave in for now for viamrtsp compatibility.

But, wasn't the whole point of having to have a memory manager the issue with passing in a pointer to the avframe in `VideoReaderFunc` return causing segfaults with the encoding in the server.

Since we are removing encoding from the server and relying on the user to handle this, shouldn't we eventually be able to get rid of the memory manager all together in viamrtsp?

Reply:
I see, interesting. I think it makes sense to leave in for now for viamrtsp compatibility.

> I think we could refactor viamrtsp though to just copy out the bytes on return early and use the new .Image pathway... I'm down to remove ReadImager entirely as long as we make it a high priority to refactor viamrtsp to use .Image and []byte

Yep the whole point to having a memory manager was the issue with passing in a pointer to the avframe in `VideoReaderFunc` return causing segfaults with the encoding in the server.

Since we are removing encoding from the server and relying on the user to handle this, we should be able to use Image pathway and simplify things quite a bit : )

Reply:
All our go cameras will need a refactor for the full fix of this interface, I do not anticipate that we will have the best version of this code and go modules after one pr. Eventually we will get rid of Stream and Next completely. The problem is those videosourcewrapper helpers are all over the place in go modules since they're exported convenience functions.

Always export code conscientiously. But we're breaking this interface anyway, so we'll deal with the following breaks.

Reply:
Okay 👍 will remove ReadImager in this PR